git/git.go

```diff
@@ -1481,3 +1481,72 @@ func ObjectDatabase(osEnv, gitEnv Environment, gitdir, tempdir string) (*gitobj.
 	}
 	return odb, nil
 }
+
+func remotesForTreeish(treeish string) []string {
+	var outp string
+	var err error
+	if treeish == "" {
+		//Treeish is empty for sparse checkout
+		tracerx.Printf("Git Treeish not provided")
+		outp, err = gitNoLFSSimple("branch", "-r", "--contains", "HEAD")
+	} else {
+		tracerx.Printf("Git Treeish %s", treeish)
+		outp, err = gitNoLFSSimple("branch", "-r", "--contains", treeish)
+	}
+	if err != nil || outp == "" {
+		tracerx.Printf("Git can't resolve symbolic name for ref: %q", treeish)
+		return []string{}
+	}
+	return strings.Split(outp, "\n")
+}
+
+// remoteForRef will try to determine the remote from the ref name.
+// This will return an empty string if any of the remote names have a slash
+// because slashes introduce ambiguity. Consider two refs:
+//
+// 1. upstream/main
+// 2. upstream/test/main
+//
+// Is the remote "upstream" or "upstream/test"? It could be either, or both.
+// We could use git for-each-ref with %(upstream:remotename) if there were a tracking branch,
+// but this is not guaranteed to exist either.
+func remoteForRef(refname string) string {
+	tracerx.Printf("Git Working Ref %", refname)
```

Comment:
I don't think this works with just a single percent sign.  Also, typically, we write the trace messages downcase, often with a prefix (so, here, “git: working ref: %s”), which would be great to fix here and elsewhere.  We typically write mostly fragments or phrases which are punctuated with colons as necessary to avoid ambiguity.

Reply:
I fixed it.
`tracerx.Printf("git: working ref: %s", refname)`
`tracerx.Printf("git: ref remote: cannot determine remote for ref %s since remote %s contains a slash", refname, name)`
`tracerx.Printf("git: working remote %s", remote)`
`tracerx.Printf("git: remote treeish: no valid remote refs parsed for %q", treeish)`
`tracerx.Printf("git: smudge: default remote failed. searching alternate remotes")`